processor/metricstransformprocessor/datapoint_aggregation.go

```diff
@@ -0,0 +1,259 @@
+	var startTimestamp *timestamp.Timestamp
+	timestampToPoints := make(map[int64][]*metricspb.Point)
+	for _, ts := range timeseries {
+		if startTimestamp == nil || ts.StartTimestamp.Seconds < startTimestamp.Seconds || (ts.StartTimestamp.Seconds == startTimestamp.Seconds && ts.StartTimestamp.Nanos < startTimestamp.Nanos) {
+			startTimestamp = ts.StartTimestamp
+		}
+		for _, p := range ts.Points {
+			if points, ok := timestampToPoints[p.Timestamp.Seconds]; ok {
+				timestampToPoints[p.Timestamp.Seconds] = append(points, p)
+			} else {
+				timestampToPoints[p.Timestamp.Seconds] = []*metricspb.Point{p}
+			}
+		}
+	}
+	return timestampToPoints, startTimestamp
+}
+
+// aggregatePoints aggregates points in the groups provided in timestampToPoints by the specific caluculation indicated by aggrType and dataType
+// Returns a group of aggregated points
+func (mtp *metricsTransformProcessor) aggregatePoints(timestampToPoints map[int64][]*metricspb.Point, aggrType AggregationType, dataType metricspb.MetricDescriptor_Type) []*metricspb.Point {
+	newPoints := make([]*metricspb.Point, 0, len(timestampToPoints))
+	for _, points := range timestampToPoints {
+		// use the timestamp from the first element because these all have the same timestamp as they are grouped by timestamp
+		timestamp := points[0].Timestamp
+		switch dataType {
+		case metricspb.MetricDescriptor_CUMULATIVE_INT64, metricspb.MetricDescriptor_GAUGE_INT64:
+			intPoint := mtp.computeInt64(points, aggrType)
+			newPoints = append(newPoints, &metricspb.Point{
+				Timestamp: timestamp,
+				Value:     intPoint,
+			})
+		case metricspb.MetricDescriptor_CUMULATIVE_DOUBLE, metricspb.MetricDescriptor_GAUGE_DOUBLE:
+			doublePoint := mtp.computeDouble(points, aggrType)
+			newPoints = append(newPoints, &metricspb.Point{
+				Timestamp: timestamp,
+				Value:     doublePoint,
+			})
+		case metricspb.MetricDescriptor_CUMULATIVE_DISTRIBUTION, metricspb.MetricDescriptor_GAUGE_DISTRIBUTION:
+			if aggrType != Sum {
+				mtp.logger.Warn("Distribution data can only be aggregated by taking the sum")
+				break
+			}
+			distPoints := mtp.computeDistribution(points)
+			for _, p := range distPoints {
+				newPoints = append(newPoints, &metricspb.Point{
+					Timestamp: timestamp,
+					Value:     p,
+				})
+			}
+
+		}
+	}
+	return newPoints
+}
+
+// computeInt64 attempts to merge int64 points into one point based on the provided aggrType
+// Returns the aggregated int64 value
+func (mtp *metricsTransformProcessor) computeInt64(points []*metricspb.Point, aggrType AggregationType) *metricspb.Point_Int64Value {
+	intVal := int64(0)
+	if len(points) > 0 {
+		intVal = points[0].GetInt64Value()
+	}
+	for _, p := range points[1:] {
+		if aggrType == Sum || aggrType == Average {
+			intVal += p.GetInt64Value()
+		} else if aggrType == Max {
+			intVal = int64(math.Max(float64(intVal), float64(p.GetInt64Value())))
```

Comment:
Can you take a max in Go without converting to / from floats? Same with min. I would think that purely integer operations would be faster than dealing with floats when not needed.

Reply:
So this is interesting, it seems like golang doesn't have a built in math function for taking max and min for integers: https://yourbasic.org/golang/max-min-function/. I used to do a manual operation for this, and then James suggested using the built-in function: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/417#discussion_r452014686. I also thought I should be using the built-in function instead of writing my own, so I changed it. Should I leave it as is or change it back to manual operation with int?

Reply:
Yea Go is missing a few helper functions 😦 

How about?
```go
intVal = max(intVal, p.GetInt64Value())

...

func max(a, b int64) int64 {
	if a < b {
		 return b;
	}

	return a;
}
```

Reply:
Yup this definitely is also my solution to go about this :) haha. It's kind of weird that go doesn't support this. I kinds want to add that in since it's an open source language too. Updated the code.